src/function.cpp

```diff
@@ -46,6 +49,31 @@ DeclaredFunction DeclaredFunction::declare(std::vector<mlir::Type> &&domain,
     throw UnsupportedException("Unsupported type: " + to_string(t));
   };
 
+  if (rangeDimRefIdx) {
+    const auto dimRefIdx = *rangeDimRefIdx;
+    smart_assert(dimRefIdx < domain.size(),
+                 "Tried to refer to the argument of an invalid index");
+
+    const auto shapedDimRef = domain[dimRefIdx].dyn_cast<mlir::ShapedType>();
+    const auto shapedRange = range.dyn_cast<mlir::ShapedType>();
+    smart_assert(
+        shapedDimRef && shapedRange,
+        "Both the specified domain and the range must be shaped types");
+    smart_assert(shapedDimRef.getRank() == shapedRange.getRank(),
+                 "The specified domain and the range must have the same ranks");
```

Comment:
Let's throw UnsupportedException in this case; assertion failure might look like a bug in mlir-tv.

Reply:
Fixed